---
title: "Guide for Users"
author: "Jingwen Yang"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette aaa}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Pacakge Installation

*AnceTran* is an *R* package that performs analyses of transcriptome evolution based on *RNA-seq* expression data or *ChIP-seq* TF binding data. 
Here, we use HNF4A-binding data for 4 mice species as an example to show how *AnceTran* works. A convenient way to install package from github is through *devtools* package:



After installation, *AnceTran* can be loaded in the usual way:




### Input Format:

*AnceTran* package takes binding score data in certain format:


- Binding score file should be a text file in the matrix shape, `Rows` correspond to orthologous. `Columns` correspond to sample names. Sample names are in format of "`TaxaName_SubtaxaName_ReplicatesName`". 

The example files are included in the AnceTran package, which can be found in `extdata` folder in the package. One can load them in to take a look:

### Construction:

The construction function `TFconstruct` loads in the BindingScore data file, and wraps them in a list of *taxonTF* objects (one *taxaTF* object).



The construction process takes **several minutes** on a desktop computer depending on data size and hardware performance. Specify **"taxa"** and **"subtaxa"** options in the function when using partial of your data. The construction process will be faster.
If you are hesitated to test the *AnceTran*, the package has already bundled a constructed object and you can load the object through:

```{r}
data(TF.objects)
```

You can take a look at what the loaded objects:

